paymentsheet-example/src/main/java/com/stripe/android/paymentsheet/example/samples/ui/server_side_confirm/ServerSideConfirmationViewModel.kt

```diff
@@ -0,0 +1,251 @@
+                )
+            }
+        }
+    }
+
+    fun handlePaymentSheetResult(paymentResult: PaymentSheetResult) {
+        val status = when (paymentResult) {
+            is PaymentSheetResult.Canceled -> null
+            is PaymentSheetResult.Completed -> "Success"
+            is PaymentSheetResult.Failed -> paymentResult.error.message
+        }
+
+        _state.update {
+            it.copy(
+                isProcessing = false,
+                status = status,
+                didComplete = paymentResult is PaymentSheetResult.Completed,
+            )
+        }
+    }
+
+    fun updateQuantity(productId: CartProduct.Id, newQuantity: Int) {
+        val dirtyCartState = state.value.cartState.updateQuantity(productId, newQuantity)
+        _state.update {
+            it.copy(dirtyCartState = dirtyCartState)
+        }
+    }
+
+    fun updateSubscription(isSubscription: Boolean) {
+        val dirtyCartState = state.value.cartState.copy(
+            isSubscription = isSubscription,
+            requiresCalculation = true,
+        )
+        _state.update {
+            it.copy(dirtyCartState = dirtyCartState)
+        }
+    }
+
+    fun handleBuyButtonPressed() {
+        _state.update {
+            it.copy(isProcessing = true)
+        }
+    }
+
+    private suspend fun prepareCheckout() {
+        val currentState = _state.updateAndGet {
+            it.copy(isProcessing = true)
+        }
+
+        val request = currentState.cartState.toCheckoutRequest()
+        val requestBody = Gson().toJson(request)
+
+        val apiResult = Fuel
+            .post("$backendUrl/checkout")
+            .jsonBody(requestBody)
+            .suspendable()
+            .awaitModel<ExampleCheckoutResponse>()
+
+        when (apiResult) {
+            is ApiResult.Success -> {
+                val newCartState = currentState.cartState.updateWithResponse(apiResult.value)
+
+                PaymentConfiguration.init(
+                    context = getApplication(),
+                    publishableKey = apiResult.value.publishableKey,
+                )
+
+                _state.update {
+                    it.copy(
+                        isProcessing = false,
+                        isError = false,
+                        confirmedCartState = newCartState,
+                        dirtyCartState = null,
+                        requiresFlowControllerConfigure = true,
+                    )
+                }
+            }
+            is ApiResult.Failure -> {
+                val status = "Preparing checkout failed\n${apiResult.error.message}"
+                _state.update {
+                    it.copy(isProcessing = false, isError = true, status = status)
+                }
+            }
+        }
+    }
+
+    @OptIn(FlowPreview::class)
+    private suspend fun observeCartUpdates() {
+        state
+            .mapNotNull { it.dirtyCartState }
+            .debounce(300.milliseconds)
```

Comment:
Is there a chance this could be out of date when we actually hit the buy button?

In our case we don't really care since the whole thing is fake. But I also don't want people to do this same thing in their apps and have it be messed up.

Reply:
Are you worried that during the debounce delay, the user could press the buy button? If so, we could move the `_state.update { it.copy(isProcessing = true) }` into an `onEach` before the `debounce` call. That would block these UI elements.

Reply:
Yes, that's the case I'm worried about!

Reply:
Fixed now!

Reply:
That’s probably because I don’t disable the quantity buttons yet when an update is in-flight. I’ll add that.